File: chatbot.js
var currentConvId = null;
var isLoading = false;

function initChatbot() {
    var messages = document.getElementById("chat-messages");
    var input = document.getElementById("chat-input");
    var sendBtn = document.getElementById("chat-send");
    var newBtn = document.getElementById("chat-new-btn");
    var exportBtn = document.getElementById("chat-export-btn");

    if (!messages) return;

    updateConvId();
    if (!messages.querySelector(".chat-bubble")) {
        showSuggestions();
    }

    if (sendBtn && input) {
        sendBtn.addEventListener("click", sendMessage);
        input.addEventListener("keydown", function (e) {
            if (e.key === "Enter" && !e.shiftKey) {
                e.preventDefault();
                sendMessage();
            }
        });
    }

    if (newBtn) {
        newBtn.addEventListener("click", newConversation);
    }

    if (exportBtn) {
        exportBtn.addEventListener("click", showExportModal);
    }

    document.querySelectorAll(".chat-suggestion-chip").forEach(function (chip) {
        chip.addEventListener("click", function () {
            if (input) {
                input.value = this.getAttribute("data-text") || this.textContent;
                sendMessage();
            }
        });
    });

    document.querySelectorAll(".chat-conv-item").forEach(function (item) {
        item.addEventListener("click", function (e) {
            if (e.target.closest(".chat-conv-delete")) return;
            loadConversation(this.getAttribute("data-conv-id"));
        });
    });

    document.querySelectorAll(".chat-conv-delete").forEach(function (btn) {
        btn.addEventListener("click", function (e) {
            e.stopPropagation();
            deleteConversation(this.getAttribute("data-conv-id"));
        });
    });

    document.querySelectorAll(".chat-export-option").forEach(function (btn) {
        btn.addEventListener("click", function () {
            exportChat(this.getAttribute("data-format"));
        });
    });

    var cancelBtn = document.querySelector(".chat-export-cancel");
    var overlay = document.querySelector(".chat-export-overlay");
    if (cancelBtn) cancelBtn.addEventListener("click", hideExportModal);
    if (overlay) overlay.addEventListener("click", hideExportModal);

    updateExportBtn();
}

function updateConvId() {
    var active = document.querySelector(".chat-conv-item.active");
    if (active) {
        currentConvId = active.getAttribute("data-conv-id");
    } else {
        var first = document.querySelector(".chat-conv-item");
        if (first) {
            currentConvId = first.getAttribute("data-conv-id");
        } else {
            currentConvId = null;
        }
    }
}

function showSuggestions() {
    var suggestions = document.getElementById("chat-suggestions");
    if (suggestions) {
        suggestions.style.display = "block";
    }
}

function hideSuggestions() {
    var suggestions = document.getElementById("chat-suggestions");
    if (suggestions) {
        suggestions.style.display = "none";
    }
}

function addBubble(text, role) {
    var messages = document.getElementById("chat-messages");
    if (!messages) return;

    var div = document.createElement("div");
    div.className = "chat-bubble " + role;

    var content = document.createElement("div");
    content.className = "chat-bubble-content";
    content.innerHTML = renderMarkdown(text);
    div.appendChild(content);

    var actions = document.createElement("div");
    actions.className = "chat-bubble-actions";
    var copyBtn = document.createElement("button");
    copyBtn.className = "chat-copy-btn";
    copyBtn.title = "Copy";
    copyBtn.innerHTML = '<i class="fas fa-copy"></i>';
    copyBtn.addEventListener("click", function () {
        copyToClipboard(text);
    });
    actions.appendChild(copyBtn);
    div.appendChild(actions);

    messages.appendChild(div);
    messages.scrollTop = messages.scrollHeight;
}

function renderMarkdown(text) {
    if (typeof marked !== "undefined" && typeof DOMPurify !== "undefined") {
        var html = marked.parse(String(text), { breaks: true, gfm: true });
        return DOMPurify.sanitize(html);
    }
    var html = String(text)
        .replace(/&/g, "&amp;")
        .replace(/</g, "&lt;")
        .replace(/>/g, "&gt;");
    html = html.replace(/###\s*(.+?)(\n|$)/g, '<h3>$1</h3>');
    html = html.replace(/##\s*(.+?)(\n|$)/g, '<h2>$1</h2>');
    html = html.replace(/#\s*(.+?)(\n|$)/g, '<h1>$1</h1>');
    html = html.replace(/\*\*(.+?)\*\*/g, '<strong>$1</strong>');
    html = html.replace(/\*(.+?)\*/g, '<em>$1</em>');
    html = html.replace(/`([^`]+)`/g, '<code>$1</code>');
    html = html.replace(/^- (.+)$/gm, '<li>$1</li>');
    html = html.replace(/((?:<li>.*?<\/li>\n?)+)/g, '<ul>$1</ul>');
    html = html.replace(/\n/g, '<br>');
    return html;
}

function sendMessage() {
    if (isLoading) return;

    var input = document.getElementById("chat-input");
    var message = input.value.trim();
    if (!message) return;

    hideSuggestions();
    addBubble(message, "user");
    input.value = "";
    setLoading(true);

    var langEl = document.documentElement;
    var language = langEl.getAttribute("data-lang") || "en";
    var districtBadge = document.querySelector(".chat-district-badge");
    var district = districtBadge ? districtBadge.textContent.replace("📍", "").trim() : "";

    fetch("/api/chat/send", {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({
            message: message,
            conversation_id: currentConvId || "",
            language: language,
            district: district,
        }),
    })
    .then(function (resp) {
        if (!resp.ok) {
            return resp.json().then(function (err) {
                throw new Error(err.message || "Server error");
            }).catch(function () {
                throw new Error("Server returned " + resp.status);
            });
        }
        return resp.json();
    })
    .then(function (data) {
        if (data.success) {
            if (data.conversation_id) {
                currentConvId = data.conversation_id;
            }
            addConversationToList(data.conversation_id, data.title);
            simulateTyping(data.reply);
        } else {
            showToast(data.message || "An error occurred", "error");
            setLoading(false);
        }
    })
    .catch(function (err) {
        setLoading(false);
        showToast(err.message || "Unable to contact the AI service at the moment. Please try again later.", "error");
    });
}

function simulateTyping(fullText) {
    var messages = document.getElementById("chat-messages");
    if (!messages) return;

    var div = document.createElement("div");
    div.className = "chat-bubble assistant";

    var content = document.createElement("div");
    content.className = "chat-bubble-content";
    content.id = "typing-content";
    div.appendChild(content);

    var actions = document.createElement("div");
    actions.className = "chat-bubble-actions";
    var copyBtn = document.createElement("button");
    copyBtn.className = "chat-copy-btn";
    copyBtn.innerHTML = '<i class="fas fa-copy"></i>';
    copyBtn.addEventListener("click", function () {
        copyToClipboard(fullText);
    });
    actions.appendChild(copyBtn);
    div.appendChild(actions);

    messages.appendChild(div);
    messages.scrollTop = messages.scrollHeight;

    var words = fullText.split(" ");
    var index = 0;
    var currentText = "";

    function typeNext() {
        if (index < words.length) {
            currentText += (index > 0 ? " " : "") + words[index];
            content.innerHTML = renderMarkdown(currentText);
            messages.scrollTop = messages.scrollHeight;
            index++;
            var delay = Math.min(50, Math.max(10, 200 / words.length));
            setTimeout(typeNext, delay);
        } else {
            content.id = "";
            setLoading(false);
            updateExportBtn();
        }
    }

    typeNext();
}

function setLoading(loading) {
    isLoading = loading;
    var sendBtn = document.getElementById("chat-send");
    var input = document.getElementById("chat-input");
    var typing = document.getElementById("chat-typing");
    if (sendBtn) sendBtn.disabled = loading;
    if (input) input.disabled = loading;
    if (typing) typing.style.display = loading ? "flex" : "none";
}

function newConversation() {
    fetch("/api/chat/new", {
        method: "POST",
        headers: { "Content-Type": "application/json" },
    })
    .then(function (r) { return r.json(); })
    .then(function (res) {
        if (res.success) {
            window.location.href = "/ai-chat?conv=" + res.conversation.id;
        }
    })
    .catch(function () {
        showToast("Failed to create new conversation", "error");
    });
}

function loadConversation(convId) {
    window.location.href = "/ai-chat?conv=" + convId;
}

function deleteConversation(convId) {
    if (!confirm("Delete this conversation?")) return;
    fetch("/api/chat/conversation/" + convId, { method: "DELETE" })
    .then(function (r) { return r.json(); })
    .then(function (res) {
        if (res.success) {
            var item = document.querySelector('.chat-conv-item[data-conv-id="' + convId + '"]');
            if (item) item.remove();
            if (currentConvId === convId) {
                window.location.href = "/ai-chat";
            }
        } else {
            showToast(res.message || "Failed to delete", "error");
        }
    })
    .catch(function () {
        showToast("Failed to delete conversation", "error");
    });
}

function addConversationToList(convId, title) {
    if (!convId) return;
    var existing = document.querySelector('.chat-conv-item[data-conv-id="' + convId + '"]');
    if (!existing) {
        var list = document.getElementById("chat-conversations");
        if (!list) return;
        var empty = list.querySelector(".chat-conv-empty");
        if (empty) empty.remove();
        var div = document.createElement("div");
        div.className = "chat-conv-item active";
        div.setAttribute("data-conv-id", convId);
        div.innerHTML =
            '<div class="chat-conv-title">' + escapeHtml(title || "New Chat") + '</div>' +
            '<div class="chat-conv-meta"><span>—</span>' +
            '<button class="chat-conv-delete" data-conv-id="' + convId + '"><i class="fas fa-trash"></i></button></div>';
        div.addEventListener("click", function () {
            loadConversation(convId);
        });
        div.querySelector(".chat-conv-delete").addEventListener("click", function (e) {
            e.stopPropagation();
            deleteConversation(convId);
        });
        list.insertBefore(div, list.firstChild);
        document.querySelectorAll(".chat-conv-item").forEach(function (item) {
            item.classList.remove("active");
        });
        div.classList.add("active");
    }
    updateExportBtn();
}

function escapeHtml(text) {
    var d = document.createElement("div");
    d.textContent = text;
    return d.innerHTML;
}

function showExportModal() {
    var modal = document.getElementById("chat-export-modal");
    if (modal) modal.style.display = "block";
}

function hideExportModal() {
    var modal = document.getElementById("chat-export-modal");
    if (modal) modal.style.display = "none";
}

function exportChat(format) {
    if (!currentConvId) return;
    fetch("/api/chat/export", {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({ conv_id: currentConvId, format: format }),
    })
    .then(function (r) { return r.json(); })
    .then(function (res) {
        if (res.success) {
            var content = res.export;
            var mimeType = res.mime;
            if (res.encoding === "base64") {
                var binary = atob(content);
                var array = new Uint8Array(binary.length);
                for (var i = 0; i < binary.length; i++) {
                    array[i] = binary.charCodeAt(i);
                }
                content = array;
            }
            var blob = new Blob([content], { type: mimeType });
            var url = URL.createObjectURL(blob);
            var a = document.createElement("a");
            a.href = url;
            a.download = res.filename;
            document.body.appendChild(a);
            a.click();
            document.body.removeChild(a);
            URL.revokeObjectURL(url);
            hideExportModal();
            showToast("Exported successfully!", "success");
        } else {
            showToast(res.message || "Export failed", "error");
        }
    })
    .catch(function () {
        showToast("Export failed", "error");
    });
}

function copyToClipboard(text) {
    if (navigator.clipboard && navigator.clipboard.writeText) {
        navigator.clipboard.writeText(text).then(function () {
            showToast("Copied!", "success");
        }).catch(function () {
            fallbackCopy(text);
        });
    } else {
        fallbackCopy(text);
    }
}

function fallbackCopy(text) {
    var ta = document.createElement("textarea");
    ta.value = text;
    ta.style.position = "fixed";
    ta.style.opacity = "0";
    document.body.appendChild(ta);
    ta.select();
    try {
        document.execCommand("copy");
        showToast("Copied!", "success");
    } catch (e) {}
    document.body.removeChild(ta);
}

function showToast(msg, type) {
    var toast = document.getElementById("chat-toast");
    if (!toast) return;
    toast.textContent = msg;
    toast.className = "chat-toast " + (type || "");
    toast.style.display = "block";
    clearTimeout(toast._hideTimer);
    toast._hideTimer = setTimeout(function () {
        toast.style.display = "none";
    }, 3000);
}

function updateExportBtn() {
    var btn = document.getElementById("chat-export-btn");
    if (!btn) return;
    var bubbles = document.querySelectorAll("#chat-messages .chat-bubble");
    btn.style.display = bubbles.length > 0 ? "flex" : "none";
}

function renderExistingMessages() {
    var contents = document.querySelectorAll("#chat-messages .chat-bubble-content");
    if (typeof marked !== "undefined" && typeof DOMPurify !== "undefined") {
        contents.forEach(function (el) {
            var raw = el.textContent;
            if (raw.indexOf("#") >= 0 || raw.indexOf("**") >= 0 || raw.indexOf("```") >= 0 || raw.indexOf("|") >= 0 || raw.indexOf("- ") >= 0) {
                el.innerHTML = DOMPurify.sanitize(marked.parse(raw, { breaks: true, gfm: true }));
            }
        });
    }
}

document.addEventListener("DOMContentLoaded", function () {
    if (document.getElementById("chat-messages")) {
        initChatbot();
        renderExistingMessages();
    }
});

File: chatbot.py
from flask import Blueprint, render_template, request, jsonify, session
from utils.auth import login_required
from services.ai_service import AIService
from models.chat import ChatConversation
from datetime import datetime
import traceback
import sys

chatbot_bp = Blueprint("chatbot", __name__)

@chatbot_bp.route("/ai-chat")
@login_required
def index():
    conv_id = request.args.get("conv", "")
    user_id = session["user_id"]
    conversations = ChatConversation.find_by_user(user_id)
    convs_data = [c.to_dict() for c in conversations]

    current_conv = None
    if conv_id:
        current_conv = ChatConversation.find_by_id(conv_id)
        if current_conv and current_conv.user_id != user_id:
            current_conv = None

    if not current_conv and convs_data:
        current_conv = conversations[0]
    elif not current_conv:
        current_conv = ChatConversation({
            "user_id": user_id,
            "title": "New Chat",
            "district": session.get("district", ""),
            "messages": [],
        })
        current_conv.save()

    return render_template(
        "ai_chat.html",
        lang=session.get("lang", "en"),
        conversations=convs_data,
        current_conv=current_conv.to_dict() if current_conv else None,
        selected_district=session.get("district", ""),
    )

@chatbot_bp.route("/api/chat/send", methods=["POST"])
@login_required
def chat_send():
    try:
        data = request.get_json()
        if not data:
            return jsonify({"success": False, "message": "Invalid request payload"}), 400

        message = data.get("message", "").strip()
        conv_id = data.get("conversation_id", "")
        language = data.get("language", session.get("lang", "en"))
        district = data.get("district", session.get("district", ""))
        user_id = session["user_id"]

        if not message:
            msg = "Please enter a question." if language == "en" else "தயவுசெய்து ஒரு கேள்வியை உள்ளிடவும்."
            return jsonify({"success": False, "message": msg})

        if len(message) < 2:
            msg = "Question too short. Please be more specific." if language == "en" else "கேள்வி மிகவும் குறுகியது. மேலும் விவரமாக கேளுங்கள்."
            return jsonify({"success": False, "message": msg})

        ai_service = AIService()

        conv = None
        history = []
        if conv_id:
            conv = ChatConversation.find_by_id(conv_id)
            if conv and conv.user_id == user_id:
                history = [{"role": m["role"], "content": m["content"]} for m in conv.messages]

        response = ai_service.get_response(
            message=message,
            language=language,
            district=district,
            history=history,
        )

        if not conv:
            conv = ChatConversation({
                "user_id": user_id,
                "title": message[:60] + ("..." if len(message) > 60 else ""),
                "district": district,
                "messages": [],
            })
            conv.save()
        elif conv.title in ("New Chat", "", None):
            title = message.strip()[:60]
            if len(message.strip()) > 60:
                title += "..."
            conv.title = title
            conv.update({"title": conv.title})

        conv.add_message("user", message)
        conv.add_message("assistant", response)

        return jsonify({
            "success": True,
            "reply": response,
            "conversation_id": conv.id,
            "title": conv.title,
        })

    except Exception as e:
        import sys
        print(f"[Chat Error] {traceback.format_exc()}", file=sys.stderr)
        lang = session.get("lang", "en")
        msg = "Unable to contact the AI service at the moment. Please try again later."
        return jsonify({"success": False, "message": msg}), 500

@chatbot_bp.route("/api/chat/conversations", methods=["GET"])
@login_required
def list_conversations():
    try:
        user_id = session["user_id"]
        conversations = ChatConversation.find_by_user(user_id)
        return jsonify({"success": True, "conversations": [c.to_dict() for c in conversations]})
    except Exception as e:
        print(f"[Chat Error] list_conversations: {traceback.format_exc()}")
        return jsonify({"success": False, "message": "Failed to load conversations"}), 500

@chatbot_bp.route("/api/chat/conversation/<conv_id>", methods=["GET"])
@login_required
def get_conversation(conv_id):
    try:
        user_id = session["user_id"]
        conv = ChatConversation.find_by_id(conv_id)
        if not conv or conv.user_id != user_id:
            lang = session.get("lang", "en")
            msg = "Conversation not found." if lang == "en" else "உரையாடல் கிடைக்கவில்லை."
            return jsonify({"success": False, "message": msg}), 404
        return jsonify({"success": True, "conversation": conv.to_dict()})
    except Exception as e:
        print(f"[Chat Error] get_conversation: {traceback.format_exc()}")
        return jsonify({"success": False, "message": "Failed to load conversation"}), 500

@chatbot_bp.route("/api/chat/conversation/<conv_id>", methods=["DELETE"])
@login_required
def delete_conversation(conv_id):
    try:
        user_id = session["user_id"]
        conv = ChatConversation.find_by_id(conv_id)
        if not conv or conv.user_id != user_id:
            lang = session.get("lang", "en")
            msg = "Conversation not found." if lang == "en" else "உரையாடல் கிடைக்கவில்லை."
            return jsonify({"success": False, "message": msg}), 404
        ChatConversation.delete_by_id(conv_id)
        return jsonify({"success": True})
    except Exception as e:
        print(f"[Chat Error] delete_conversation: {traceback.format_exc()}")
        return jsonify({"success": False, "message": "Failed to delete conversation"}), 500

@chatbot_bp.route("/api/chat/new", methods=["POST"])
@login_required
def new_conversation():
    try:
        user_id = session["user_id"]
        district = session.get("district", "")
        conv = ChatConversation({
            "user_id": user_id,
            "title": "New Chat",
            "district": district,
            "messages": [],
        })
        conv.save()
        return jsonify({"success": True, "conversation": conv.to_dict()})
    except Exception as e:
        print(f"[Chat Error] new_conversation: {traceback.format_exc()}")
        return jsonify({"success": False, "message": "Failed to create conversation"}), 500

@chatbot_bp.route("/api/chat/export", methods=["POST"])
@login_required
def export_chat():
    try:
        data = request.get_json()
        conv_id = data.get("conv_id", "")
        export_format = data.get("format", "txt")
        lang = session.get("lang", "en")
        user_id = session["user_id"]

        conv = ChatConversation.find_by_id(conv_id)
        if not conv or conv.user_id != user_id:
            msg = "Conversation not found." if lang == "en" else "உரையாடல் கிடைக்கவில்லை."
            return jsonify({"success": False, "message": msg}), 404

        now = datetime.utcnow().strftime("%Y-%m-%d %H:%M")
        lang_label = "Tamil" if lang == "ta" else "English"

        if export_format == "pdf":
            from fpdf import FPDF
            from fpdf.enums import XPos, YPos
            pdf = FPDF()
            pdf.add_page()
            pdf.set_auto_page_break(auto=True, margin=15)
            pdf.add_font("Arial", "", r"C:\Windows\Fonts\arial.ttf")
            pdf.add_font("Arial", "B", r"C:\Windows\Fonts\arialbd.ttf")
            pdf.set_font("Arial", "B", 16)
            pdf.cell(0, 10, text="AI Agriculture Assistant - Chat Export", new_x=XPos.LMARGIN, new_y=YPos.NEXT, align="C")
            pdf.set_font("Arial", "", 10)
            pdf.cell(0, 6, text=f"Date: {now}", new_x=XPos.LMARGIN, new_y=YPos.NEXT)
            pdf.cell(0, 6, text=f"District: {conv.district or 'Not set'}", new_x=XPos.LMARGIN, new_y=YPos.NEXT)
            pdf.cell(0, 6, text=f"Language: {lang_label}", new_x=XPos.LMARGIN, new_y=YPos.NEXT)
            pdf.line(10, pdf.get_y() + 2, 200, pdf.get_y() + 2)
            pdf.ln(6)

            pdf.set_font("Arial", "", 11)
            for msg in conv.messages:
                role = "You" if msg["role"] == "user" else "AI Assistant"
                pdf.set_x(pdf.l_margin)
                pdf.set_font("Arial", "B", 11)
                pdf.multi_cell(0, 6, text=f"[{role}]")
                pdf.set_x(pdf.l_margin)
                pdf.set_font("Arial", "", 11)
                pdf.multi_cell(0, 6, text=msg.get("content", ""))
                pdf.set_x(pdf.l_margin)
                pdf.ln(3)

            pdf_output = bytes(pdf.output())
            import base64
            return jsonify({
                "success": True,
                "export": base64.b64encode(pdf_output).decode("ascii"),
                "filename": f"chat_{conv_id[:8]}.pdf",
                "mime": "application/pdf",
                "encoding": "base64",
            })
        else:
            lines = []
            header = (
                f"AI Agriculture Assistant - Chat Export\n"
                f"Date: {now}\n"
                f"District: {conv.district or 'Not set'}\n"
                f"Language: {lang_label}\n"
                f"{'=' * 50}\n\n"
            )
            lines.append(header)

            for msg in conv.messages:
                role = "You" if msg["role"] == "user" else "AI Assistant"
                lines.append(f"[{role}]\n{msg['content']}\n\n")

            return jsonify({
                "success": True,
                "export": "".join(lines),
                "filename": f"chat_{conv_id[:8]}.txt",
                "mime": "text/plain",
            })
    except Exception as e:
        print(f"[Chat Error] export: {traceback.format_exc()}")
        return jsonify({"success": False, "message": "Failed to export chat"}), 500
